lib/clientv3/ippool_kdd_conversion_test.go

```diff
@@ -0,0 +1,249 @@
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/ipip"
+	"github.com/projectcalico/libcalico-go/lib/options"
+	"github.com/projectcalico/libcalico-go/lib/testutils"
+	"github.com/projectcalico/libcalico-go/lib/watch"
+)
+
+var _ = testutils.E2eDatastoreDescribe("IPPool KDD v1 to v3 migration tests", testutils.DatastoreK8s, func(config apiconfig.CalicoAPIConfig) {
+
+	ctx := context.Background()
+	name1 := "ippool-1"
+	name2 := "ippool-2"
+	spec1_v1 := apiv3.IPPoolSpec{
+		CIDR:        "1.2.3.0/24",
+		NATOutgoing: true,
+		IPIP: &apiv1.IPIPConfiguration{
+			Enabled: true,
+			Mode:    ipip.CrossSubnet,
+		},
+	}
+	spec1_v3 := apiv3.IPPoolSpec{
+		CIDR:        "1.2.3.0/24",
+		NATOutgoing: true,
+		IPIPMode:    apiv3.IPIPModeCrossSubnet,
+	}
+	kvp1_v3 := &model.KVPair{
+		Key: model.ResourceKey{
+			Name: "1.2.3.0-24",
+			Kind: apiv3.KindIPPool,
+		},
+		Value: &apiv3.IPPool{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       apiv3.KindIPPool,
+				APIVersion: apiv3.GroupVersionCurrent,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "1-2-3-0-24",
+			},
+			Spec: apiv3.IPPoolSpec{
+				CIDR:     "1.2.3.0/24",
+				Disabled: false,
+				IPIP: &apiv1.IPIPConfiguration{
+					Enabled: true,
+					Mode:    ipip.CrossSubnet,
+				},
+			},
+		},
+	}
+
+	spec2_v1 := apiv3.IPPoolSpec{
+		CIDR:        "2001::/120",
+		NATOutgoing: true,
+		IPIP: &apiv1.IPIPConfiguration{
+			Enabled: false,
+		},
+	}
+	spec2_v3 := apiv3.IPPoolSpec{
+		CIDR:        "2001::/120",
+		NATOutgoing: true,
+		IPIPMode:    apiv3.IPIPModeNever,
+	}
+	kvp2_v3 := &model.KVPair{
+		Key: model.ResourceKey{
+			Name: "2001---120",
+			Kind: apiv3.KindIPPool,
+		},
+		Value: &apiv3.IPPool{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       apiv3.KindIPPool,
+				APIVersion: apiv3.GroupVersionCurrent,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "2001---120",
+			},
+			Spec: apiv3.IPPoolSpec{
+				CIDR:     "2001::/120",
+				Disabled: false,
+				IPIP: &apiv1.IPIPConfiguration{
+					Enabled: false,
+				},
+			},
+		},
+	}
+	DescribeTable("IPPool CRD with v1 IPIP field tests",
+		func(name1, name2 string, spec_v1, spec_v3 apiv3.IPPoolSpec, kvp *model.KVPair) {
+			c, err := clientv3.New(config)
+			Expect(err).NotTo(HaveOccurred())
+
+			be, err := backend.NewClient(config)
+			Expect(err).NotTo(HaveOccurred())
+			be.Clean()
+
+			By("Attempting to creating a new IPPool with the non-writable v1 IPIP field")
+			_, outError := c.IPPools().Create(ctx, &apiv3.IPPool{
```

Comment:
I assume here you could have just use kvp.Value.(*apiv3.IPPool) rather than needing to pass in the spec_v1

Reply:
it also feels more of a validation that going through the backend with the same resource works.